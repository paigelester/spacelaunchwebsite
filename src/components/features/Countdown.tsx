import React, { useState, useEffect } from 'react';
import moment from 'moment';

type CountdownTimeRemaining = {
    days: string;
    hours: string;
    minutes: string;
    seconds: string;
};

interface CountdownProps {
    targetDate: string;
}

const Countdown = (props: CountdownProps) => {

    const countdownFunc = (): CountdownTimeRemaining => {
        const now: any = moment();
        const target: any = moment(props.targetDate);

        const countdown = moment(target - now);

        return {
            days: countdown.format("D"),
            hours: countdown.format("HH"),
            minutes: countdown.format("mm"),
            seconds: countdown.format("ss")
        };
    };

    const [timeRemaining, setTimeRemaining] = useState<CountdownTimeRemaining>(countdownFunc());

    useEffect(() => {
        const countdownInterval = setInterval(() => setTimeRemaining(countdownFunc), 1000);

        return () => clearInterval(countdownInterval);
    }, [countdownFunc]);

    return (
        <div className="countdown">
            <div className="text-center me-2 d-inline-block">
                <p className="countdown-days-value fs-1 m-0">{timeRemaining.days}</p>
                <p className="countdown-days text-muted text-uppercase m-0">days</p>
            </div>
            <div className="text-center me-2 d-inline-block">
                <p className="countdown-hours-value fs-1 m-0">{timeRemaining.hours}</p>
                <p className="countdown-hours text-muted text-uppercase m-0">hours</p>
            </div>
            <div className="text-center me-2 d-inline-block">
                <p className="countdown-minutes-value fs-1 m-0">{timeRemaining.minutes}</p>
                <p className="countdown-minutes text-muted text-uppercase m-0">minutes</p>
            </div>
            <div className="text-center d-inline-block">
                <p className="countdown-seconds-value fs-1 m-0">{timeRemaining.seconds}</p>
                <p className="countdown-seconds text-muted text-uppercase m-0">seconds</p>
            </div>
        </div>
    );
};

export default Countdown;