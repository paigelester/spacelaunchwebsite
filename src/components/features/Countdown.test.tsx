import React from 'react';
import { render } from '@testing-library/react';
import moment from 'moment';
import Countdown from './Countdown';


beforeEach(() => {
    Date.now = jest.fn(() => 1618171860892);
});

// it("renders as expected", () => {
//     // given
//     let targetDate: Date = new Date(Date.now());
//     targetDate.setDate(targetDate.getDate() + 1);
//     targetDate.setSeconds(targetDate.getTime() + 1);
//     targetDate.setMinutes(targetDate.getTime() + 1);
//     targetDate.setHours(targetDate.getTime() + 1);

//     // when
//     const { container } = render(<Countdown targetDate={targetDate.toString()} />);

//     // then
//     expect(container).toMatchSnapshot();
// });

it("renders the number of days away from target date", () => {
    // given
    let targetDate = moment().add(1, "days");

    // when
    const { container } = render(<Countdown targetDate={targetDate.toString()} />);

    const daysText = container.querySelector(".countdown-days");
    const daysValue = container.querySelector(".countdown-days-value");

    // then
    expect(daysText?.textContent).toContain("days");
    expect(daysValue?.textContent).toContain("2");
});

// it("renders the number of hours away from target date", () => {
//     // given
//     let targetDate = new Date(Date.now());
//     targetDate.setSeconds(targetDate.getTime() + 1);

//     // when
//     const { container } = render(<Countdown targetDate={targetDate.toString()} />);

//     const hoursText = container.querySelector(".countdown-hours");
//     const hoursValue = container.querySelector(".countdown-hours-value");

//     // then
//     expect(hoursText?.textContent).toContain("hours");
//     expect(hoursValue?.textContent).toContain("0");
// });

// it("renders the number of minutes away from target date", () => {
//     // given
//     let targetDate = new Date(Date.now());
//     targetDate.setSeconds(targetDate.getTime() + 1);

//     // when
//     const { container } = render(<Countdown targetDate={targetDate.toString()} />);

//     const minutesText = container.querySelector(".countdown-minutes");
//     const minutesValue = container.querySelector(".countdown-minutes-value");

//     // then
//     expect(minutesText?.textContent).toContain("minutes");
//     expect(minutesValue?.textContent).toContain("34");
// });

// it("renders the number of seconds away from target date", () => {
//     // given
//     let targetDate = new Date(Date.now());
//     targetDate.setSeconds(targetDate.getTime() + 1);

//     // when
//     const { container } = render(<Countdown targetDate={targetDate.toString()} />);

//     const secondsText = container.querySelector(".countdown-seconds");
//     const secondsValue = container.querySelector(".countdown-seconds-value");

//     // then
//     expect(secondsText?.textContent).toContain("seconds");
//     expect(secondsValue?.textContent).toContain("52");
// });

// it("should countdown after 1 second", () => {
//     // given
//     let targetDate = new Date(Date.now());
//     targetDate.setSeconds(targetDate.getTime() + 1);

//     // when
//     const { container } = render(<Countdown targetDate={targetDate.toString()} />);

//     const secondsValue = container.querySelector(".countdown-seconds-value");

//     // then
//     setTimeout(() => {
//         expect(secondsValue?.textContent).toContain("52");
//     }, 1000);
// });
