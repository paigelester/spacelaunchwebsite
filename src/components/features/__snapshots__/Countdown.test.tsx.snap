// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders as expected 1`] = `
<div>
  <div
    class="countdown"
  >
    <div
      class="text-center me-2 d-inline-block"
    >
      <p
        class="countdown-days-value fs-1 m-0"
      >
        NaN
      </p>
      <p
        class="countdown-days text-muted text-uppercase m-0"
      >
        days
      </p>
    </div>
    <div
      class="text-center me-2 d-inline-block"
    >
      <p
        class="countdown-hours-value fs-1 m-0"
      >
        NaN
      </p>
      <p
        class="countdown-hours text-muted text-uppercase m-0"
      >
        hours
      </p>
    </div>
    <div
      class="text-center me-2 d-inline-block"
    >
      <p
        class="countdown-minutes-value fs-1 m-0"
      >
        NaN
      </p>
      <p
        class="countdown-minutes text-muted text-uppercase m-0"
      >
        minutes
      </p>
    </div>
    <div
      class="text-center d-inline-block"
    >
      <p
        class="countdown-seconds-value fs-1 m-0"
      >
        NaN
      </p>
      <p
        class="countdown-seconds text-muted text-uppercase m-0"
      >
        seconds
      </p>
    </div>
  </div>
</div>
`;
